### ISSUE TITLE
Allow optional quorum on votes

### ISSUE THREAD

**Issue by ethanfrey** [iov-one/weave#539]
**Is your feature request related to a problem? Please describe.**
After discussing the election design, we came up with a few more features we need to define. This is a placeholder for them.

**Describe the solution you'd like**

Allow election rules to specify a quorum (what percent must vote) as well as a threshold (what percent of the voting parties must vote yes). Current work is a threshold of yes votes out of *all possible voters*. The quorum version is often used where voting is not mandatory. Counting out of total possible votes is used in parliments. We may want to enable both versions (eg. Quorum not set -> threshold of all possible voters) 

Also: ensure abstain vote properly works with both quorum and non-quorum election rules. This indicates participation and helps reach quorum if you participate but don't really care which side wins

**Comment by alpe** [comment]
I did some redesign to support the quorum by a threshold (th) value in election rules. By calculating a quorum and threshold weight value from the electorate, the proposal acceptance logic can be achieved by the following steps.
* calculating weights (on proposal creation)
```go
	var quorumThresholdWeight, acceptanceThresholdWeight uint64
	if quorum != nil {
		quorumThresholdWeight = totalElectorateWeight * uint64(quorum.Numerator) / uint64(quorum.Denominator)
		acceptanceThresholdWeight = quorumThresholdWeight * uint64(threshold.Numerator) / uint64(threshold.Denominator)
	} else {
		quorumThresholdWeight = totalElectorateWeight * uint64(threshold.Numerator) / uint64(threshold.Denominator)
		acceptanceThresholdWeight = quorumThresholdWeight
	}

	// prevent threshold weights that can never be fulfilled
	if quorumThresholdWeight >= totalElectorateWeight {
		quorumThresholdWeight = totalElectorateWeight - 1
	}
	if acceptanceThresholdWeight >= totalElectorateWeight {
		acceptanceThresholdWeight = totalElectorateWeight - 1
	}
	return TallyResult{
		QuorumThresholdWeight:     quorumThresholdWeight,
		acceptanceThresholdWeight: acceptanceThresholdWeight,
	}

```
* Calculate both weights exceeded (on tally)
```
	total := m.TotalVotes()
	return total > m.acceptanceThresholdWeight && uint64(m.TotalYes) > m.QuorumThresholdWeight
```

❗️  The only problem I see is that without further constraints on the values we can achieve a situation where a proposal gets accepted even with a majority of NO votes. See quorum 1/2, threshold 1/2  .


## Number examples
electorate weight| quorum th fraction| acceptance th fraction| => | quorum th weight | acceptance th weight| yes weight| no weight| abstain weight| result| comment
----------|-------------------|-----------------------|----|------------------|---------------------|----|---|--------|-------|---
9|  - |  1/2|       |  4|  4|  5|  0|  0|  accepted | quorum and acceptance thresholds exceeded
9|  - |  1/2|       |  4|  4|  4|  0|  0|  rejected | quorum threshold not exceeded 
9|  - |  1/2|       |  4|  4|  4|  1|  0|  rejected | acceptance threshold not exceeded 
9|  - |  1/2|       |  4|  4|  4|  5|  0|  rejected | acceptance threshold not exceeded
9|  - |  1/2|       |  4|  4|  4|  0|  1|  rejected | acceptance threshold not exceeded
9|  - |  2/3|       |  6|  6|  6|  0|  0|  rejected | quorum threshold not exceeded
9|  - |  1/1|		|  8|  8|  9|  0|  0|  accepted | quorum and acceptance thresholds exceeded
9| 1/2|  1/2|       |  4|  2|  5|  0|  0|  accepted | quorum and acceptance thresholds exceeded 
9| 1/2|  1/2|       |  4|  2|  4|  0|  0|  rejected | quorum threshold not exceeded 
9| 1/2|  1/2|       |  4|  2|  4|  1|  0|  accepted | quorum and acceptance thresholds exceeded 
🚨 9| 1/2|  1/2|       |  4|  2|  4|  5|  0|  **accepted** | quorum and acceptance thresholds exceeded 
9| 1/2|  1/2|       |  4|  2|  4|  0|  1|  accepted | quorum and acceptance thresholds exceeded 
9| 1/2|  1/2|       |  4|  2|  3|  1|  1|  accepted | quorum and acceptance thresholds exceeded 
9| 1/2|  1/2|       |  4|  2|  2|  2|  1|  rejected | acceptance thresholds not exceeded 
9| 1/2|  2/3|       |  4|  2|  3|  1|  1|  accepted | quorum and acceptance thresholds exceeded 
9| 2/3|  1/2|       |  6|  3|  3|  0|  1|  rejected | quorum threshold not exceeded 
9| 2/3|  1/2|       |  6|  3|  3|  0|  4|  rejected | acceptance threshold not exceeded 
9| 2/3|  1/2|       |  6|  3|  7|  0|  0|  accepted | quorum and acceptance thresholds exceeded 
9| 2/3|  1/2|       |  6|  3|  6|  0|  1|  accepted | quorum and acceptance thresholds exceeded 
9| 2/3|  2/3|       |  6|  4|  6|  0|  1|  accepted | quorum and acceptance thresholds exceeded 
9| 1/1|  1/1|       |  8|  8|  9|  0|  0|  accepted | quorum and acceptance thresholds exceeded

**Comment by alpe** [comment]
🙈 I did not read it correct. The threshold of Yes votes is off the voting parties. Which makes sense so that with an electorate of 10, quorum 1/2, threshold 1/2 the following example applies:

votes| Yes | accepted
------|-----|------
5      |  5   | no (quorum not exceeded)
6      |     3 |  no
6     |     4 |  yes
7    |    4  | yes
8  |      4|  no
8  |   5 | yes

**Comment by ethanfrey** [comment]
Nice table. I should have done something like that for the issue beforehand.
I guess that is what grooming sessions are supposed to shake out.

**Comment by ethanfrey** [comment]
Was closed by #574
